Use system installed deflate library

bket opened the pull request:
Use pkgconfig if available to locate a libdeflate system library.

If a specific library prefix is given by LIBDEFLATE_PREFIX=somepath, use that.
Otherwise if pkgconfig is installed and the library can be located via that, use that one.
Otherwise (or if enforced by USE_BUNDLED_DEFLATE=yes), fall back to bundled code.

Build tested, with and without deflate library installed, on OpenBSD
amd64 current. Tests pass.

ThomasWaldmann commented on setup.py:
```diff
@@ -13,24 +15,62 @@
+        installed = default
+        try:
+            installed = pkgconfig_installed(lib, required_version)
+        except EnvironmentError:
+            # no pkg-config present
+            pass
+        except ValueError:
+            # pkgconfig was unable to determine if required version of
+            # libdeflate is available. Bundled version of libdeflate will be
+            # used.
+            pass
```
I guess one would want to emit some warning, so the user will know what's missing / how to fix that.

ThomasWaldmann replied:
it also could emit a info msg when it has found libdeflate via pkg-config and will use that.

ThomasWaldmann replied:
hmm, if you always fall back to bundled code if pkg-config did not work, it's maybe rather a info msg in all cases.
just tell the user whether it used the pkg-config located system lib or the bundled stuff because ...

bket replied:
I have a new diff, which mimics the build behavior of borgbackup. This new diff tries to use pkconfig when available. A user can force usage of the bundeld library by having `USE_BUNDLED_DEFLATE` in the environment. Additionally a user can set `LIBDEFLATE_PREFIX` to workaround not having pkgconfig on the system. I will force push this diff in a bit.

ThomasWaldmann commented on setup.py:
```diff
@@ -13,24 +15,62 @@
+
+extension_kwargs = {}
+
+if libdeflate_found is True:
```
```
if libdeflate_found:
```

bket replied:
Yes, you are right.

ThomasWaldmann commented on setup.py:
```diff
@@ -62,4 +102,7 @@
         'Topic :: System :: Archiving :: Compression',
     ],
     python_requires='>=3.6',
+    setup_requires=[
+        "pkgconfig",
+    ],
```
did you try how that behaves?

i tried this for `borgbackup`, but the problem is that `setup.py` already needs `pkgconfig`  before `setup()` is executed (and could install `pkgconfig` due to `setup_requires`).

bket replied:
You are right again, this does not work. This has been addressed.

ThomasWaldmann commented on setup.py:
```diff
@@ -13,24 +15,62 @@
+    # pkgconfig is not installed. It will be installed by setup_requires.
+    pass
+else:
+    def pkgconfig_installed_check(lib, required_version, default):
```
what would be the point of `default=True`?

bket replied:
I left it there by accident. Removed

ThomasWaldmann commented on setup.py:
```diff
@@ -4,6 +4,8 @@
 
 from setuptools import Extension, setup
 
+DEFLATE_REQUIRED_VERSION = ">= 1.10"
+
```
you could add something to force using the bundled code (default: try finding and using the system lib), see how borgbackup deals with the libs in `setup*.py`.

https://github.com/borgbackup/borg/blob/1.2.0/setup.py#L49

bket replied:
Done :-)

ThomasWaldmann reviewed (commented):
nice addition, i made something similar back then for borg's bundled code / libs.

bket committed:
Use system installed deflate library

Check if a suitable deflate library is installed on the system using the
pkgconfig package. Alternatively a user can set `LIBDEFLATE_PREFIX` in
their environment, e.g. `LIBDEFLATE_PREFIX=/usr/local`. Force usage of
bundled version by setting `USE_BUNDLED_DEFLATE` in environment.

Based on feedback from @ThomasWaldmann.

ThomasWaldmann commented on setup.py:
```diff
@@ -13,24 +13,52 @@
     version = re.match(r'.*#define MODULE_VERSION "(.*?)"', src.read(), re.S).group(1)
 
-cpu = "arm" if platform.machine().startswith("arm") else "x86"
+prefer_system_libdeflate = not bool(os.environ.get("USE_BUNDLED_DEFLATE"))
```
this is True by default (if the env var is not used).

ThomasWaldmann commented on setup.py:
```diff
@@ -13,24 +13,52 @@
+]
+
+if prefer_system_libdeflate:
+    if system_prefix_libdeflate:
+        print("Detected and preferring libdeflate [via LIBDEFLATE_PREFIX]")
+        extension_kwargs = dict(include_dirs=[os.path.join(system_prefix_libdeflate, "include")],
+                                library_dirs=[os.path.join(system_prefix_libdeflate, "lib")],
+                                libraries=["deflate"])
+
+    elif pc and pc.installed("libdeflate", ">= 1.10"):
+        print("Detected and preferring libdeflate [via pkgconfig]")
+        extension_kwargs = pc.parse("libdeflate")
```
this is the block executed by default.

there's a problem now if there is no pkgconfig (pc is None) and the user did not set LIBDEFLATE_PREFIX (which is also usually the default).

so, people who do not read the docs and install the lib and pkgconfig before this, will get a build failure - **not a fallback to bundled code**.

bket replied:
Good catch, should be resolved now.

ThomasWaldmann reviewed (changes requested):
good improvement, but there is a popular failure mode now instead of a fallback.

ThomasWaldmann commented:
also minor pep8 issue:
```
setup.py:55:21: missing whitespace around operator
```

ThomasWaldmann commented on setup.py:
```diff
@@ -13,24 +13,51 @@
+    print("Detected and preferring libdeflate [via pkgconfig]")
+    extension_kwargs = pc.parse("libdeflate")
+else:
+    print("Using bundeld libdeflate")
```
typo

ThomasWaldmann replied:
same typo also in commit comment.

bket replied:
Fixed

ThomasWaldmann commented:
BTW, I edited the top post a bit to reflect the current behaviour.

Could also just be a c&p of the commit comment.

ThomasWaldmann commented:
BTW, our tests of course only test the default behaviour and do not set any of these env vars yet - but I suspect you tried all modes manually?

bket commented:
Fixed :-)

bket commented:
> BTW, our tests of course only test the default behaviour and do not set any of these env vars yet - but I suspect you tried all modes manually?

Yes, tried all modes manually. So far, behavioud is as expected.

ThomasWaldmann reviewed (approved):
LGTM!

ThomasWaldmann commented:
@dcwatson I think this is a good and very flexible improvement to `deflate` and it especially enables dist package maintainers to not use the bundled code (or even remove it) if they rather prefer to use the `libdeflate` library from another dist package.

Avoids duplication and eases maintenance, especially in case of critical bug or security fixes, they then only need to get fixed at one place.

Merge?

dcwatson commented on setup.py:
```diff
@@ -13,24 +13,51 @@
     version = re.match(r'.*#define MODULE_VERSION "(.*?)"', src.read(), re.S).group(1)
 
-cpu = "arm" if platform.machine().startswith("arm") else "x86"
+prefer_system_libdeflate = not bool(os.environ.get("USE_BUNDLED_DEFLATE"))
```
`USE_BUNDLED_DEFLATE=0` will use the bundled libdeflate, which seems counterintuitive, especially if something is trying to set that env to a true or false value based on a condition (rather than setting/not setting). I think having a list of true or false values and some documentation will take care of this.

ThomasWaldmann replied:
that's basically my quick and dirty code @bket took from borgbackup.

we only support not setting it or setting it to =yes (or in fact =anything, which all evals to true).

i agree that true / false / True / False / yes / no / 0 / 1 (anything else?) would be cleaner.

ThomasWaldmann replied:
made it more clean/flexible now.

dcwatson commented on setup.py:
```diff
@@ -13,24 +13,51 @@
+try:
+    import pkgconfig as pc
+except ImportError:
+    print("Warning: can not import pkgconfig python package.")
```
Do we need this warning if using the bundled version?

ThomasWaldmann replied:
No, when forcing the bundled code, no pkgconfig is used.

ThomasWaldmann replied:
no warning any more if not used anyway.

ThomasWaldmann commented:
@bket If you don't mind, I'll fix it now according to the review comments of @dcwatson ...

bket commented:
> @bket If you don't mind, I'll fix it now according to the review comments of @dcwatson ...

@ThomasWaldmann, thank you for taking care of this! Changes lgtm, and make sense.

ThomasWaldmann commented:
@dcwatson have a look now!

ThomasWaldmann commented:
@bket if this gets merged and released, it will unblock some borg PRs, that's why I am trying to speed this up a bit. :-)

ThomasWaldmann commented:
maybe squash all commits together instead of merging all the separate commits.

ThomasWaldmann committed:
cibuildwheel: automatically run pytest

ThomasWaldmann commented:
trying pytest within cibuildwheel now vvv.

ThomasWaldmann committed:
disable building pypy wheels

Problem:
  ImportError while importing test module '/project/tests/test_compress.py'.
  Hint: make sure your test modules/packages have valid Python names.
  Traceback:
  /opt/python/pp37-pypy37_pp73/lib-python/3/importlib/__init__.py:127: in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
  /project/tests/test_compress.py:6: in <module>
      import deflate
  E   ImportError: /tmp/tmp.Dh9bFArfRD/venv/site-packages/deflate.pypy37-pp73-x86_64-linux-gnu.so: undefined symbol: Py_Initialize

ThomasWaldmann commented:
yay! all except pypy tested OK.

Guess pypy breakage is not new and unrelated to this PR?
```
  ImportError while importing test module '/project/tests/test_compress.py'.
  Hint: make sure your test modules/packages have valid Python names.
  Traceback:
  /opt/python/pp37-pypy37_pp73/lib-python/3/importlib/__init__.py:127: in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
  /project/tests/test_compress.py:6: in <module>
      import deflate
  E   ImportError: /tmp/tmp.Dh9bFArfRD/venv/site-packages/deflate.pypy37-pp73-x86_64-linux-gnu.so: undefined symbol: Py_Initialize
```

ThomasWaldmann commented:
@dcwatson have a look! before merging i need to remove that temporary commit.

https://github.com/dcwatson/deflate/runs/5489155309?check_suite_focus=true

ThomasWaldmann commented:
@dcwatson considering this works and wheels are even tested now, i'ld like to merge this if you don't object.

this will remove the pypy wheel build (which likely was broken before also, just nobody noticed).

i'll remove that temporary commit now, which was just for demonstration purposes, but we still have the test results linked above.

ThomasWaldmann commented:
temporary commit removed, rebased on current master.

ThomasWaldmann commented:
As there were no more objections, I merged this now.

In case we find more to do / to fix, it can be done in a later PR.

Next thing we need is a new release. @dcwatson can you do that or give me permissions on pypi.org to do that? #26
